Added FakeWebAssembleHostEnvironment, tests, and docs. (#856)
bUnit-dev/bUnit: docs/site/docs/test-doubles/fake-webassemblyhostenvironment.md

---
uid: fake-webassemblyhostenvironment
title: Faking IWebAssemblyHostEnvironment
---

# Faking `IWebAssemblyHostEnvironment`

bUnit has a fake implementation of Blazor's `IWebAssemblyHostEnvironment` built-in, which is added by default to bUnit's `TestContext.Services` service provider. That means nothing special is needed to test components that depend on `IWebAssemblyHostEnvironment`, as it is already available by default.

> 

Reviewer: ```suggestion

Out of the box, the fake implementation has its `Environment` property set to `production`, and its `BaseAddress` set to `/`.

```

> ## Verify `IWebAssemblyHostEnvironment` interactions
> 
> Lets look at a few examples that show how to verify that a component correctly uses the `IWebAssemblyHostEnvironment` in various ways.

Reviewer: ```suggestion
## Setting `Environment` and `BaseAddress`

Lets look at a few examples of how to set the `IWebAssemblyHostEnvironment` two properties `Environment` and `BaseAddress` via the built-in fake.
```

Reviewer: Its a bit pedantic, but we're not really verifying whether the user code is correctly using  IWebAssemblyHostEnvironment.  We're simply setting a state in one, and then observing the result after. There is no guarantee that users actually did call the IsDevelopment method, for that we would need a mock.

In the examples, we'll use the following `<HelloWorld>` component:

```cshtml
@inject IWebAssemblyHostEnvironment HostEnvironment

> <p>Hello @(HostEnvironment.IsDevelopment() ? "Developers" : "World"). The base URL is: @HostEnvironment.BaseAddress</p>

Reviewer: ```suggestion
<p id="message">
  Hello @(HostEnvironment.IsDevelopment() ? "Developers" : "World"). 
</p>
<p id="address">
  The base URL is: @HostEnvironment.BaseAddress
</p>
```

```

To verify that the `<HelloWorld>` component correctly says hello to the developers, do the following:

```csharp
// Arrange
using var ctx = new TestContext();
var hostEnvironment = ctx.Services.GetRequiredService<FakeWebAssemblyHostEnvironment>();

> hostEnvironment.SetEnvironmentToDevelopment();

Reviewer: ```suggestion

// Sets the environment to "Development". There are two other helper 
// methods available as well, SetEnvironmentToProduction() and 
// set SetEnvironmentToStaging(), or environment can also be changed
// directly through the hostEnvironment.Environment property.
hostEnvironment.SetEnvironmentToDevelopment();
```

> var cut = ctx.RenderComponent<SimpleUsingWebAssemblyHostEnvironment>();

Reviewer: ```suggestion
// Act
var cut = ctx.RenderComponent<SimpleUsingWebAssemblyHostEnvironment>();
```

> // Assert - inspects markup to verify the message
> cut.Find("p").MarkupMatches($"<p>Hello Developers. The base URL is: /</p>");

Reviewer: ```suggestion
// Assert - inspects markup to verify the message
cut.Find("#message").MarkupMatches($"<p>Hello Developers.</p>");
```

```

To verify that the `<HelloWorld>` component correctly uses the current `BaseAddress`, do the following:

```csharp
// Arrange
using var ctx = new TestContext();
var hostEnvironment = ctx.Services.GetRequiredService<FakeWebAssemblyHostEnvironment>();

> hostEnvironment.BaseAddress = "myBaseUrl/";
> var cut = ctx.RenderComponent<SimpleUsingWebAssemblyHostEnvironment>();
> 
> // Assert - inspect markup to verify that the BaseAddress is used correctly.
> cut.Find("p").MarkupMatches($"<p>Hello World. The base URL is: myBaseUrl/</p>");

Reviewer: ```suggestion
// Sets a new base address directly on the BaseAddress property.
hostEnvironment.BaseAddress = "myBaseUrl/";

// Act
var cut = ctx.RenderComponent<SimpleUsingWebAssemblyHostEnvironment>();

// Assert - inspect markup to verify that the BaseAddress is used correctly.
cut.Find("#address").MarkupMatches($"<p>The base URL is: myBaseUrl/</p>");
```
